Common/Securities/CurrencyConversion/SecurityCurrencyConversion.cs

```diff
@@ -0,0 +1,268 @@
+        {
+            SourceCurrency = sourceCurrency;
+            DestinationCurrency = destinationCurrency;
+
+            _steps = steps;
+        }
+
+        /// <summary>
+        /// Updates the internal conversion rate based on the latest data, and returns the new conversion rate
+        /// </summary>
+        /// <returns>The new conversion rate</returns>
+        public decimal Update()
+        {
+            var newConversionRate = 0m;
+            var stepWithDataFound = false;
+
+            _steps.ForEach(step =>
+            {
+                var lastData = step.RateSecurity.GetLastData();
+                if (lastData == null)
+                {
+                    return;
+                }
+
+                var price = lastData.Price;
+                if (price == 0m)
+                {
+                    return;
+                }
+
+                if (!stepWithDataFound)
+                {
+                    newConversionRate = 1m;
+                    stepWithDataFound = true;
+                }
+
+                if (step.Inverted)
+                {
+                    newConversionRate /= price;
+                }
+                else
+                {
+                    newConversionRate *= price;
+                }
+            });
+
+            ConversionRate = newConversionRate;
+            return ConversionRate;
+        }
+
+        /// <summary>
+        /// Finds a conversion between two currencies by looking through all available 1 and 2-leg options
+        /// </summary>
+        /// <param name="sourceCurrency">The currency to convert from</param>
+        /// <param name="destinationCurrency">The currency to convert to</param>
+        /// <param name="existingSecurities">The securities which are already added to the algorithm</param>
+        /// <param name="potentialSymbols">The symbols to consider, may overlap with existingSecurities</param>
+        /// <param name="makeNewSecurity">The function to call when a symbol becomes part of the conversion, must return the security that will provide price data about the symbol</param>
+        /// <returns>A new <see cref="SecurityCurrencyConversion"/> instance representing the conversion from sourceCurrency to destinationCurrency</returns>
+        /// <exception cref="ArgumentException">Thrown when no conversion from sourceCurrency to destinationCurrency can be found</exception>
+        public static SecurityCurrencyConversion LinearSearch(
+            string sourceCurrency,
+            string destinationCurrency,
+            IList<Security> existingSecurities,
+            IEnumerable<Symbol> potentialSymbols,
+            Func<Symbol, Security> makeNewSecurity)
+        {
+            var allSymbols = existingSecurities.Select(sec => sec.Symbol).Concat(potentialSymbols)
+                .Where(x => x.SecurityType == SecurityType.Crypto || x.Value.Length == 6)
```

Comment:
> x.Value.Length == 6

hm this looks fishy?

Believe we already filter in the caller for `.Where(s => s.Type == SecurityType.Forex || s.Type == SecurityType.Cfd || s.Type == SecurityType.Crypto)` ? `potentialSymbols` is created using forex/cfd and crypto too

We could filter `allSymbols` for `if (!potentialConversionRateSymbol.PairContainsCurrency(sourceCurrency))` once I think here, and avoid doing it on each loop bellow, note that the `ToList()` call here is already looping over all of this once